src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -0,0 +1,371 @@
+        LOGGER.error(
+            "Failed to convert broken native record to byte data:\n{}",
+            serializeError.getMessage());
+        throw e;
+      }
+    }
+    // create new sinkRecord
+    Schema keySchema = isKey ? new SnowflakeJsonSchema() : record.keySchema();
+    Object keyContent = isKey ? newSFContent : record.key();
+    Schema valueSchema = isKey ? record.valueSchema() : new SnowflakeJsonSchema();
+    Object valueContent = isKey ? record.value() : newSFContent;
+    return new SinkRecord(
+        record.topic(),
+        record.kafkaPartition(),
+        keySchema,
+        keyContent,
+        valueSchema,
+        valueContent,
+        record.kafkaOffset(),
+        record.timestamp(),
+        record.timestampType(),
+        record.headers());
+  }
+
+  /**
+   * Invokes insertRows API using the offsets buffered for this partition. TODO: SNOW-530369 Retry
+   * handling mechanisms
+   */
+  public void insertBufferedRows() {
+    PartitionBuffer intermediateBuffer = null;
+    bufferLock.lock();
+    try {
+      intermediateBuffer = streamingBuffer;
+      streamingBuffer = new StreamingBuffer();
+    } finally {
+      // release lock
+      bufferLock.unlock();
+    }
+    // intermediate buffer can be empty here if time interval reached but kafka produced no records.
+    if (intermediateBuffer.isEmpty()) {
+      LOGGER.info("No Rows Buffered, returning;");
+      this.previousFlushTimeStampMs = System.currentTimeMillis();
+      return;
+    }
+    InsertValidationResponse response =
+        channel.insertRows(
+            (Iterable<Map<String, Object>>) intermediateBuffer.getData(),
+            Long.toString(intermediateBuffer.getLastOffset()));
+
+    // need to update the flush time(last I called insertrows)
+    this.previousFlushTimeStampMs = System.currentTimeMillis();
+
+    // should we retry if response.hasErrors() == true?
+    // how can we retry if buffer was cleared?
+    // lets have an exponential back off and jitter retry policy
+
+    if (response.hasErrors()) {
+      LOGGER.error(
+          "Inserted rows API response has errors for rows with offset start:{} and end:{}",
+          intermediateBuffer.getFirstOffset(),
+          intermediateBuffer.getLastOffset());
+      response
+          .getInsertErrors()
+          .forEach(
+              insertError -> {
+                LOGGER.error(
+                    "Insert row Error message:{}, with ex:{}",
+                    insertError.getMessage(),
+                    insertError.getException().getMessage());
+              });
+    }
+  }
+
+  // should we rely on precommits frequency to give us the last committed offset token?
+  // or should we have a BG thread polling on every channel to update the Atomic long and simply
+  // return the atomic long's current value
+  // For now we will rely on preCommit API to fetch the inserted offsets into Snowflake
+  // TODO: SNOW-529755 POLL committed offsets in backgraound thread
```

Comment:
Is there any correctness issue if we reply on precommit frequency?

Reply:
Not really, precommit is called immediately after put (in the same poll thread). Also even if there are not messages, we still received a call to put with 0 messages which means precommit will be called.  

offset.flush.interval.ms is defaulted to 60 seconds. And this can be set at connect level. 

Also, it is not necessary we need to commit data in precommit. It is very much possible that 
- put(1 record) -> puts in buffer
- precommit(1 record) we still return 0 since insertrows was not called. 

But in subsequent put APIs calls which will again call precommit will fetch offsetToken 1 since insertRows would have been called ultimately. 


There is a potential problem that flush set by customers is < offset.flush.interval.ms, in this case latency will be high since flush time would have passed but we could only know that after `offset.flush.interval.ms`

Reply:
For context here is kafka code: https://github.com/apache/kafka/blob/830d83e2cde4f33735128b533cacc8b8ea23a2c0/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java#L210